scapy/contrib/automotive/bmw/hsfz.py

```diff
@@ -97,8 +97,13 @@ def send(self, x):
         except AttributeError:
             pass
 
-        return super(UDS_HSFZSocket, self).send(
-            HSFZ(src=self.src, dst=self.dst) / x)
+        try:
+            return super(UDS_HSFZSocket, self).send(
+                HSFZ(src=self.src, dst=self.dst) / x)
+        except Exception as e:
+            log_interactive.error("Exception: %s", e)
+            self.close()
+            return 0
```

Comment:
Thanks for asking... this requires a little bit of background information. Automotive ECUs which support HSFZ or DoIP usually have multiple different "States" and usually at least two interfaces, CAN and Ethernet. On state is the normal firmware, another state is for example the bootloader. Since the bootloader has a much smaller code base, than the firmware, in most cases, Ethernet (DoIP or HSFZ) is not supported. This means, as soon as an ECU enters the bootloader state, the Ethernet interface goes down. 

This has some drawbacks for an automated scan, I want to do with UDS_Scanner. During a scan, the scanner will find a bootloader state and by default will try to run all enumerators against this bootloader to check which services are supported. The scanner will send an ECU to bootloader state, an then try to execute the enumerator. The enumerator will execute this line: https://github.com/secdev/scapy/blob/0421836e53447400668119edce8ae8843d6c678f/scapy/contrib/automotive/scanner/enumerator.py#L174-L176
Since, in the meantime the ECU put's its interface down, an Exception during send (_sndrcv_snd) will come up. Here is a very general catch: https://github.com/secdev/scapy/blob/23ff728153479378f388aa244525e45ccd679b8e/scapy/sendrecv.py#L251-L252
This catch now hides the exception from my enumerator. So the enumerator has now way to figure out, if the interface of the ECU went down. The enumerator will receive a None, and will try the next packet. This means the enumerator will fire a huge number of requests against a dead interface which floods the output with exception log messages. 

I didn't want to change the code in _sndrcv_snd, therefore I did the exception handling inside this custom sockets. In this way, the socket goes down and the enumerator will know, that the socket died.

Reply:
I think that you should add some comments to the code =)